### PR TITLE
@tus/s3-store: Fix incorrect part retreival procedure

### PR DESCRIPTION
**Pull request by mitjap**
This PR fixes some issues with `retrieveParts` function.

First issue is how it determines when to fetch more parts. Before it was using `data.NextPartNumberMarker && Number(data.NextPartNumberMarker) > 0` to check if there are more parts to be fetch. This code made an assumption that final "page" will return "0" as `NextPartNumberMarker` but this is not specified in documentation and differs between S3 providers. Scaleway's S3 for example always return last part's number. Specification is clear on this that we should use `IsTruncated` propery.

Next issue is with sorting the parts. When completing multipart upload parts must be ordered in ascending order. In existing code sorting was done only when there less then 1000 parts (no additional "pages").

Another issue is with filtering. I'm not sure why would we ever need to remove parts from this list.

### CODE REVIEW

File: packages/s3-store/index.ts [hunk 1]
```diff
@@ -490,7 +478,8 @@ export class S3Store extends DataStore {
     // Metadata request needs to happen first
     const metadata = await this.getMetadata(id)
     const parts = await this.retrieveParts(id)
-    const partNumber = parts?.length ?? 0
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const partNumber: number = parts.length > 0 ? parts[parts.length - 1].PartNumber! : 0
     const nextPartNumber = partNumber + 1
```
Comment (mitjap) [inline packages/s3-store/index.ts:483]: `nextPartNumber` should be calculated from last part number

File: packages/s3-store/index.ts [hunk 2]
```diff
@@ -378,38 +378,26 @@ export class S3Store extends DataStore {
...
+      const rest = await this.retrieveParts(id, data.NextPartNumberMarker)
+      parts = [...parts, ...rest]
+    }
 
-    if (parts && !partNumberMarker) {
-      return (
-        parts
-          // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-          .sort((a, b) => a.PartNumber! - b.PartNumber!)
-          .filter((value, index) => value.PartNumber === index + 1)
```
Comment (mitjap) [inline packages/s3-store/index.ts:411]: Maybe this filter was kind of fail-safe if some part in the middle was not uploaded. This does not prevent later file corruption if some part was uploaded only partially.

**Review by Murderlon — APPROVED** [review]
Makes sense!

**Comment by mitjap** [comment]
@Murderlon Would be possible to publish this fix?

**Comment by Murderlon** [comment]
Done!
https://github.com/tus/tus-node-server/releases/tag/%40tus%2Fs3-store%401.1.1

**Pull request merged**